// admin-firebase/src/FirebaseManager.ts
import { ref } from 'firebase/storage';


export interface FireBaseModules {
  database: any;
  firestore: any;
  functions: any;
  remoteConfig: any;
  auth: any;
  storage: any;
}


class FirebaseManager {
  private static instance: FirebaseManager;
  private modules?: FireBaseModules;
  private storagePath = 'images';
  public root : string | null = null;

  private constructor() { }

  public static getInstance(): FirebaseManager {
    if (!FirebaseManager.instance) {
      FirebaseManager.instance = new FirebaseManager();
    }
    return FirebaseManager.instance;
  }

  public setModules(modules: FireBaseModules): void {
    this.modules = modules;
  }

  // Lazily retrieve the auth module, for example
  public get auth() {
    if (!this.modules || !this.modules.auth) {
      throw new Error('Firebase auth module is not initialized.');
    }
    return this.modules.auth;
  }

  // Similarly, add getters for other modules like firestore, database, etc.
  public get firestore() {
    if (!this.modules || !this.modules.firestore) {
      throw new Error('Firebase firestore module is not initialized.');
    }
    return this.modules.firestore;
  }

  public get functions() {
    if (!this.modules || !this.modules.functions) {
      throw new Error('Firebase functions module is not initialized.');
    }
    return this.modules.functions;
  }

  public get remoteConfig() {
    if (!this.modules || !this.modules.remoteConfig) {
      throw new Error('Firebase remoteConfig module is not initialized.');
    }
    return this.modules.remoteConfig;
  }

  public get database() {
    if (!this.modules || !this.modules.database) {
      throw new Error('Firebase database module is not initialized.');
    }
    return this.modules.database;
  }

  public get storage() {
    if (!this.modules || !this.modules.storage) {
      throw new Error('Firebase storage module is not initialized.');
    }
    return this.modules.storage;
  }

  // ... add similar getters for other modules ...

  public getStorageRef(path: string | null = null): any {
    if (!this.modules || !this.modules.storage) {
      throw new Error('Firebase storage module is not initialized.');
    }
    if (path === null) {
      path = this.storagePath;
    }
    return ref(this.modules.storage, path);
  }

  public getModules(): FireBaseModules {
    if (!this.modules) {
      throw new Error('Firebase modules are not initialized.');
    }
    return {
      database: this.database,
      firestore: this.firestore,
      functions: this.functions,
      remoteConfig: this.remoteConfig,
      auth: this.auth,
      storage: this.storage
    }
  }
  public setRoot(root: string | null): void {
    this.root = root;
  }
}


export default FirebaseManager
