// admin-firebase/src/models/index.ts

import { Image, ModelMeta, customClaims,TrackItem } from './interfaces';

export abstract class BaseModel {
    static readonly META: ModelMeta = {
        foreignKeys: [],
        icon: "mdi-account",
        includeInDashboard: false
    };

    public propertyTypes = new Map<string, string>();
    public propertyModels = new Map<string, string>();
    constructor(public key: string,) {
        this.key = key
    }
    public propertyPermissions = new Map<string, Array<string>>();
    public classPermission = new  Map<string, Array<string>>();
    public readonly serviceType: string = "base"
    toTable(): Array<Object> {
        const list = [];
        var modelArgs = null
        for (const prop of Object.getOwnPropertyNames(this)) {
            if (prop == "META" || prop == "foreignKeys" || prop == "propertyTypes" || prop == "propertyModels" || prop == "serviceType" || prop == "propertyPermissions" || prop == "classPermission" || prop.includes('option')) continue;
            let foreign = false;
            for (const foreignKey of this.getMeta().foreignKeys) {
                if (foreignKey == prop) {
                    foreign = true;
                    break;
                }
            }
            let hide = false;
            // if (prop == "key") hide = true;
            if (this.getPropertyType(prop) == "Options") {
                //@ts-ignore
                modelArgs = this[prop + '_options']
            }
            list.push({ name: prop, label: prop.charAt(0).toUpperCase() + prop.slice(1), field: prop, align: 'center', foreign: foreign, modeltype: this.getPropertyType(prop), dbmodel: this.getPropertyModel(prop), hide: hide, modelargs: modelArgs });
        }
        return list;
    }
    // toObject():Object{
    //     if (this.key==undefined) {
    //         let {key, ...obj}=Object.assign({}, this);
    //         return obj;
    //     };
    //     return Object.assign({}, this);
    // }
    toObject() {
        const result: any = {};

        for (const prop of Object.getOwnPropertyNames(this)) {
            if (prop == "meta" || prop == "propertyTypes" || prop == "serviceType" || prop == "propertyPermissions" || prop == "classPermission" || prop == "propertyModels") continue;

            let foreign = false;
            for (const foreignKey of this.getMeta().foreignKeys) {
                if (foreignKey == prop) {
                    foreign = true;
                    //check if value is array
                    break;
                }
            }

            if (this.getProp(prop) == undefined) {
                result[prop] = null;
            } else if (foreign) {
                if (Array.isArray(this.getProp(prop))) {
                    result[prop] = this.getProp(prop).map((item:any) => item.key || item);
                } else {
                    result[prop] = this.getProp(prop).key || this.getProp(prop);
                }
            } else {
                result[prop] = this.getProp(prop);
            }
        }

        return result;
    }
    getProp(prop: string): any {
        return (this as any)[prop];
    }
    getPropertyType(propertyName: string): string|undefined {
        if (this.propertyTypes.has(propertyName)) {
            return this.propertyTypes.get(propertyName);
        }
        return 'undefined';
    }
    getPropertyModel(propertyName: string): string|undefined {
        if (this.propertyModels.has(propertyName)) {
            return this.propertyModels.get(propertyName);
        }
        return 'string';
    }
    getMeta(): ModelMeta {
        return Object.getPrototypeOf(this).constructor.META;
    }
    fields(): Array<Object> {
        const list = [];
        for (const prop of Object.getOwnPropertyNames(this)) {
            if (prop == "meta" || prop == "propertyTypes" || prop == "serviceType" || prop == "propertyPermissions" || prop == "classPermission" || prop == "propertyModels") continue;
            let foreign = false;
            for (const foreignKey of this.getMeta().foreignKeys) {
                if (foreignKey == prop) {
                    foreign = true;
                    break;
                }
            }
            if (this.getProp(prop) == undefined) {
                var value = null;
            }
            else if (foreign) {
                // console.log(prop)
                var value = this.getProp(prop).map((item: any) => item.key);
            } else {
                var value = this.getProp(prop);
            }
            // console.log(foreign)
            let hide = false;
            // if (prop == "key") hide = true;
            list.push({ name: prop, label: prop.charAt(0).toUpperCase() + prop.slice(1), value: value, foreign: foreign, hide: hide, type: typeof value, modeltype: this.getPropertyType(prop), dbmodel: this.getPropertyModel(prop) });
        }
        return list;
    }

    getClassName(): string {
        return this.constructor.name;
    }
    toCASL(role: string): Object {
        var rawRules = []
        var subject = this.getClassName()
        for (const [key, value] of this.classPermission) {
            if (key == role) {
                for (const val of value) {

                    let item = {
                        action: val, subject: subject
                    }
                    rawRules.push(item)
                }
            }

        }
        for (const [property, permissions] of this.propertyPermissions) {
            for (const [key, value] of permissions) {
                let item = {
                    actions: value,
                    subject: `${subject}.${property}`
                }
                rawRules.push(item)
            }
        }
        return rawRules
    }
    representation(): string {
        return this.key
    }

    fromObject(obj: any) {
        for (const prop of Object.getOwnPropertyNames(this)) {
            if (prop == "meta" || prop == "propertyTypes" || prop == "serviceType" || prop == "propertyPermissions" || prop == "classPermission" || prop == "propertyModels") continue;
            if (obj[prop] == undefined) {
                (this as any)[prop] = null;
            } else {
                (this as any)[prop] = obj[prop];
            }
        }
    }

}



export class User extends BaseModel {
    static readonly META: ModelMeta = {
        foreignKeys: [],
        icon: "mdi-account",
        includeInDashboard: false
    };
    constructor(public key: string, public displayName: string, public email: string, public emailVerified: boolean, public photoURL: string, public customClaims: customClaims = { role: "admin", age: 0,tracks:[],locale:'hu' }) {
        super(key);
        this.displayName = displayName;
        this.email = email;
        // this.password = password;
        this.photoURL = photoURL;
        this.emailVerified = emailVerified;
        this.customClaims = customClaims;

        this.propertyTypes.set('key', 'String');
        this.propertyTypes.set('displayName', 'String');
        this.propertyTypes.set('email', 'String');
        //this.propertyTypes.set('password', 'String');
        this.propertyTypes.set('photoURL', 'Image');
        this.propertyTypes.set('emailVerified', 'Boolean');
        this.propertyTypes.set('customClaims', 'Interface');

        this.classPermission.set('admin', ['write', 'read', 'delete'])
        this.classPermission.set('organizer', ['write', 'read'])

    }
}




export type ModelConstructors = {
    [key: string]: new (...args: any[]) => any;
  };
  
  class ModelRegistry {
    private static instance: ModelRegistry;
  
    private models: ModelConstructors = {
      // You can initialize default models here
      'User':User
    };
  
    private constructor() {}
  
    public static getInstance(): ModelRegistry {
      if (!ModelRegistry.instance) {
        ModelRegistry.instance = new ModelRegistry();
      }
      return ModelRegistry.instance;
    }
  
    registerModel(name: string, model: new (...args: any[]) => any) {
    //   console.log(name)
      this.models[name] = model;
    }
  
    getModels(): ModelConstructors {
      return this.models;
    }
  }

  
  export default ModelRegistry;
