// admin-firebase/src/services/admin.ts

import ModelRegistry from '../models'
import FirebaseManager from '../FirebaseManager';

const models = ModelRegistry.getInstance().getModels()

const firebaseModules = FirebaseManager.getInstance().getModules();

export class AdminService {

    get(id: string) {
        return 
    }
    getAll() {
        return 
    }
    create(data: any) {
        return
    }
    update(data: any) {
        return []
    }
    delete(id: string) {
        return
    }
    loadAllToModel() {
        return []
    }

    async login(email: string) {
        return firebaseModules.functions.httpsCallable("loginGuest")({ email });
    }
    async register(email: string, fullName: string) {
        return firebaseModules.functions.httpsCallable("registerNewGuest")({ fullName, email })
    }
    async retrieveModel(snapshot: any) {
        var modelobj = new models['User']()
        var modelFields = modelobj.fields()
        for (let item in modelFields) {
            var field = modelFields[item]
            modelobj[field.name] = snapshot[field.name] ? snapshot[field.name] : field.value;
        }
        modelobj.key = snapshot.uid
        modelobj.id = snapshot.uid
        return modelobj
    }
    async loadToModel(id: string) {
        var modelobj:any = undefined

        await this.getUsers([id]).then(async (snapshot: any) => {
            console.log(snapshot.data)
            if (snapshot.length == 0) return modelobj;
            snapshot = snapshot.data[0]
            console.log(snapshot)
            modelobj = await this.retrieveModel(snapshot)
        }).catch((error: any) => {
            console.log(error)
        })
        return modelobj
    }

    async getUsers(filter: Array<string> = []) {
        return firebaseModules.functions.httpsCallable("getUsers")({ filter });
    }
    async checkIfEmailExists(email: string) {
        return firebaseModules.functions.httpsCallable("checkIfEmailExists")({ email });
    }
    async verifyEmail(email: string, password: string) {
        return firebaseModules.functions.httpsCallable("verifyEmail")({ email, password });
    }
    //create user
    async createUser(uid: any) {
        return firebaseModules.functions.httpsCallable("createUser")({ uid });
    }
    //update user
    async updateUser(uid: any, data: any) {
        console.log(data)
        if (data.photoURL == '') {
            data.photoURL = null
        }
        if (typeof data.customClaims == 'string') {
            data.customClaims = JSON.parse(data.customClaims)
        }
        return firebaseModules.functions.httpsCallable("updateUser")({ uid, data });
    }
    //delete user
    async deleteUser(uid: any) {
        return firebaseModules.functions.httpsCallable("deleteUser")({ uid });
    }
    //create entity
    async createEntity(data: any) {
        return firebaseModules.functions.httpsCallable("createProduct")(data);
    }
    //update entity
    async updateEntity(data: any) {
        return firebaseModules.functions.httpsCallable("updateProduct")(data);
    }
    //delete entity
    async deleteEntity(data: any) {
        return firebaseModules.functions.httpsCallable("deleteProduct")(data);
    }
    async getRemoteConfig(value: string) {
        var result = null
        await firebaseModules.remoteConfig.fetchAndActivate()
            .then(() => {
                result = firebaseModules.remoteConfig.getValue(value);
                // console.log(`Welcome message: ${welcomeMessage.asString()}`);
            })
            .catch((error: any) => {
                console.error(`Error fetching remote config: ${error}`);
            });
        return result
    }

}
