// admin-firebase/src/services/auth.ts
import FirebaseManager from "../FirebaseManager";

const firebaseModules = FirebaseManager.getInstance().getModules();
// const auth = FirebaseManager.getInstance().auth;

class AuthService {


  async login(email: string, password: string) {
    return firebaseModules.auth.signInWithEmailAndPassword(email.trim(), password);
  }
  async register(email: string, password: string) {
    return firebaseModules.auth.createUserWithEmailAndPassword(email, password)
  }
  async logout() {
    return firebaseModules.auth.signOut()
  }
  async currentUser() {
    return firebaseModules.auth.currentUser
  }
  async getCustomClaims(): Promise<any> {
    return new Promise((resolve, reject) => {
      this.currentUser().then((user) => {
        if (user) {
          console.log(user)
          user.reload()
          user.getIdTokenResult(true)
            .then((idTokenResult: any) => {
              console.log(idTokenResult.claims);
              resolve(idTokenResult.claims);
            })
            .catch((error: any) => {
              console.error(error);
              reject(error);
            });
        } else {
          reject('User not logged in');
        }
      }).catch((error) => {
        console.error(error);
        reject(error);
      });
    });
  }

}

export default AuthService
