// admin-firebase/src/services/base.ts
import ModelRegistry from "../models";
import FirebaseManager from '../FirebaseManager';
import { AdminService } from "./admin";
import firebase from 'firebase/compat';
import 'firebase/firestore';
import { FireBaseModules } from "../FirebaseManager";
type DocumentSnapshot = firebase.firestore.DocumentSnapshot;
type QuerySnapshot = firebase.firestore.QuerySnapshot;
const firebaseModules = FirebaseManager.getInstance().getModules();

const models = ModelRegistry.getInstance().getModels()

interface QueryInterface {
  attribute: string;
  relation: string;
  refval: any;
  foreign?: boolean
}

function getValueByPath(obj: object, path: string): any {
  const pathParts = path.split('.');
  let currentValue: any = obj;

  for (const part of pathParts) {
    if (currentValue.hasOwnProperty(part)) {
      currentValue = currentValue[part];
    } else {
      return undefined;
    }
  }

  return currentValue;
}


function applyFilter(model: any, params: QueryInterface): boolean {
  var modelAttr = getValueByPath(model, params.attribute);
  if (params.foreign) {
    modelAttr = modelAttr?.key
  }
  switch (params.relation) {
    case '==':
      return modelAttr === params.refval;
    case '>':
      return modelAttr > params.refval;
    case '<':
      return modelAttr < params.refval;
    case '>=':
      return modelAttr >= params.refval;
    case '<=':
      return modelAttr <= params.refval;
    default:
      return false;
  }
}


class BaseService {
  db: any;
  constructor(public model: string, public root: string|null = null) {
    if (this.root !== null) {
      this.db = firebaseModules.database.ref(`/${this.root}/${model.toLowerCase()}`);
    }else{
      this.db = firebaseModules.database.ref(`/${model.toLowerCase()}`);
    }
  }
  async get(id: string) {
    return this.db.child(id).once("value");
  }
  async getAll() {
    return this.db.once("value");
  }
  async create(data: any) {
    // console.log(data)
    const ref = this.db.push(data);
    return ref
  }
  async update(id: string, data: any,newData:boolean=false) {
    if (newData){
      return this.db.set(data);

    }
    return this.db.child(id).update(data);
  }
  async delete(id: string) {
    // console.log("delete", id)
    return this.db.child(id).remove();
  }

  async retrieveModel(snapshot: any) {
    var bs:AdminService|Service
    //  console.log(snapshot)
    var modelobj = new models[this.model]();
    var modelFields = modelobj.fields();
    for (let item in modelFields) {
      var field = modelFields[item]
      // console.log(field.name)
      // console.log(modelobj[field.name])
      if (field.foreign) {
        if (Array.isArray(field.value)) {
          modelobj[field.name] = []
          for (let item in snapshot[field.name]) {
            var foreignKey = snapshot[field.name][item]
            // console.log(field.name)
            if (field.modeltype != 'User') {
               bs = new Service(field.label)
              // console.log(foreignKey)
            } else {
               bs = new AdminService()
            }
            // console.log(foreignKey)
            var newm = await bs.loadToModel(foreignKey)
            modelobj[field.name].push(newm)
            // console.log(modelobj)
          }
        } else if (field.value !== null || field.value !== undefined) {
          var foreignKey = snapshot[field.name]
          if (field.modeltype != 'User') {
             bs = new Service(field.dbmodel)
            // console.log(foreignKey)
          } else {
             bs = new AdminService()
          }
          var newm = await bs.loadToModel(foreignKey)
          // console.log(newm)
          modelobj[field.name] = newm
        } else {
          modelobj[field.name] = null
        }


      } else {
        // BUGGY
        modelobj[field.name] = snapshot[field.name] ? snapshot[field.name] : field.value;

        //resolved
        //modelobj[field.name]=snapshot[field.name] if (snapshot[field.name]!=undefined); else field.value;
      }
    }
    return modelobj
  }
  async loadToModel(id: string) {
    var modelobj: any = undefined
    // // console.log(id)
    await this.get(id).then(async (snapshot: any) => {
      // // console.log(snapshot.val())
      if (snapshot.val() == null) return modelobj;
      var { uid, ...attr } = snapshot.val();
      //   // console.log(attr)
      modelobj = await this.retrieveModel(attr);
    });
    if (modelobj == undefined) return modelobj;
    return modelobj;
  }
  async loadAllToModel(snapshot: any | null = null) {
    if (snapshot == null) {
      snapshot = await this.getAll()
    }
    var modelobj = undefined
    var modelarr = []
    // console.log(snapshot.val())
    for (let childSnapshot in snapshot.val()) {
      var { uid, ...attr } = snapshot.val()[childSnapshot];
      modelobj = await this.retrieveModel(attr);
      modelobj.key = childSnapshot
      modelarr.push(modelobj)
    }
    return modelarr;
  }
}



class FireService {
  db: any;
  constructor(public model: string, public root: string|null = null) {
    this.model = model
    if (this.root !== null) {
      this.db = firebaseModules.firestore.collection(`wedding/${this.root}/${model.toLowerCase()}`);
    } else {
      this.db = firebaseModules.firestore.collection(model.toLowerCase());
    }
  }
  async get(id: string) {
    return this.db.doc(id).get();
  }
  async getAll() {
    return this.db.get();
  }
  async create(data: any) {
    const filteredData = Object.entries(data)
      .filter(([key, value]) => {
        return (
          typeof value === 'string' ||
          typeof value === 'number' ||
          typeof value === 'boolean' ||
          value instanceof Date ||
          Array.isArray(value)
        );
      })
      .reduce((obj:any, [key, value]) => {
        obj[key] = value;
        return obj;
      }, {});

    // Add filtered data to Firestore
    const docRef = await this.db.add(filteredData);
    const doc = await docRef.get();
    return { id: doc.id, ...doc.data() };


  }
  async update(id: string, data: any,newData:boolean=false) {
    // Convert the data object to an array of [key, value] pairs
    const dataEntries = Object.entries(data);

    // Loop through the data entries and remove any custom Map objects
    const sanitizedData:any = {};
    for (const [key, value] of dataEntries) {
      if (!(value instanceof Map)) {
        sanitizedData[key] = value;
      }
    }
    // console.log(sanitizedData)
    // Update the document with the sanitized data
    if (newData){
      return this.db.doc(id).set(sanitizedData);
    }
    return this.db.doc(id).update(sanitizedData);
  }

  async delete(id: string) {
    return this.db.doc(id).delete();
  }
  async retrieveModel(doc: any) {
    var bs:AdminService|Service
    var modelobj = new models[this.model]();
    var modelFields = modelobj.fields();
    for (let item in modelFields) {
      var field = modelFields[item]
      var snapshotval = doc.data()?.[field.name]
      if (field.foreign) {
        if (Array.isArray(field.value)) {
          modelobj[field.name] = []
          for (const foreignKey of doc.data()?.[field.name] || []) {
            if (field.modeltype != 'User') {
               bs = new Service(field.label)
              // console.log(foreignKey)
            } else {
               bs = new AdminService()
            }
            // console.log(foreignKey)
            const newm = await bs.loadToModel(foreignKey);
            modelobj[field.name].push(newm);
          }
        } else if (snapshotval !== null && snapshotval !== undefined) {
          // var foreignKey = doc.data?.[field.name]
          if (field.modeltype != 'User') {
             bs = new Service(field.dbmodel)
            // console.log(snapshotval)
          } else {
             bs = new AdminService()
          }
          // console.log(snapshotval)
          // console.log(typeof snapshotval)
          var newm = await bs.loadToModel(snapshotval)
          // console.log(newm)
          modelobj[field.name] = newm
        } else {
          modelobj[field.name] = null
        }

      } else {
        modelobj[field.name] = doc.data()?.[field.name] ?? field.value;
      }
    }
    return modelobj;
  }
  async loadToModel(id: string) {
    const doc = await this.db.doc(id).get();
    if (!doc.exists) return undefined;
    const modelobj = await this.retrieveModel(doc);
    modelobj.key = doc.id;
    return modelobj;
  }
  async loadAllToModel(snapshot: QuerySnapshot | null = null) {
    // console.log(this.model)
    if (!snapshot) {
      snapshot = await this.getAll();
    } else {
      snapshot = await snapshot
    }

    // console.log(snapshot)
    var modelarr = [];
    if (snapshot !== null){
      for (const doc of snapshot.docs) {
        // console.log(snapshot)
        // console.log(doc)
        const modelobj = await this.retrieveModel(doc);
        modelobj.key = doc.id;
        modelarr.push(modelobj);
      };
    }
    return modelarr;
  }

}
class Service {

  public service: FireService | BaseService | AdminService;
  db: any
  root: string | null = FirebaseManager.getInstance().root;
  constructor(public model: string) {
    if (model == "User") {
      this.service = new AdminService() 
      this.db = null
    } else {
      // console.log(models[model])
      var modelObj = new models[model]();
      var fire = modelObj.serviceType == "fire"
      // console.log(modelObj)
      // console.log(fire)
      this.service = fire ? new FireService(model,this.root) : new BaseService(model,this.root)
      // this.service=new BaseService(model)
      this.db = this.service.db
    }
  }

  // Define the methods that you want to expose in the Service class
  async get(id: string) {
    return this.service.get(id);
  }

  async getAll() {
    return this.service.getAll();
  }

  async create(data: any) {
    return this.service.create(data);
  }

  async update(id: string, data: any,newData:boolean=false) {
    return this.service.update(id, data,newData);
  }

  async delete(id: string) {
    return this.service.delete(id);
  }
  async retrieveModel(id: any) {
    return this.service.retrieveModel(id)
  }

  async loadToModel(id: string) {
    return this.service.loadToModel(id);
  }

  async loadAllToModel(snapshot: any) {
    return this.service.loadAllToModel(snapshot);
  }
  async query(queryparams: QueryInterface[], func: string) {
    // console.log(attribute, func, relation, refval);
    const self = this;
    try {
      const models:any[] = await self.service.loadAllToModel();
      var filteredModels = models
      if (queryparams.length > 0) {
        filteredModels = models.filter((model: any) => {
          return queryparams.every((queryParam) => {
            return applyFilter(model, queryParam);
          });
        });
      }
      let result: number | any[];

      switch (func) {
        case 'sum':
          result = filteredModels.reduce((a: number, b: any) => a + getValueByPath(b, queryparams[queryparams.length - 1].attribute), 0);
          break;
        case 'average':
          const sum = filteredModels.reduce((a: number, b: any) => a + getValueByPath(b, queryparams[queryparams.length - 1].attribute), 0);
          result = sum / filteredModels.length;
          break;
        case 'count':
          result = filteredModels.length;
          break;
        case 'filter':
          result = filteredModels
          break;
        default:
          return new Error('Invalid function, supported functions are sum, average, count');
      }

      return result;
    } catch (error) {
      console.error('Error querying Firestore:', error);
      throw error;
    }
  }
}


export { BaseService, FireService, Service, type QueryInterface, type FireBaseModules };
