// admin-firebase/index.ts
// Export Models
export * as models from './src/models';

// Export Services
// export * as services from './src/services';

// Export Components
export { default as AdminLayout } from './src/layouts/AdminLayout.vue';
export { default as Uploader } from './src/components/Uploader.vue';
export { default as DashboardSumItem } from './src/components/DashboardSumItem.vue';
export { default as DashboardFormItems } from './src/components/DashboardFormItem.vue';
export { default as DataLoadMixin } from './src/components/DataLoadMixin';
export { default as DataModelMixin } from './src/components/DataModelMixin';
export { default as ForeignKey } from './src/components/ForeignKey.vue';
export { default as ModelCell } from './src/components/ModelCell.vue';
export { default as ModelForm } from './src/components/ModelForm.vue';
export { default as ModelTable } from './src/components/ModelTable.vue';
export { default as AdminBoolean } from './src/components/tools/AdminBoolean.vue';
export { default as AdminImage } from './src/components/tools/AdminImage.vue';
export { default as AdminImages } from './src/components/tools/AdminImages.vue';
export { default as AdminInterface } from './src/components/tools/AdminInterface.vue';
// export { default as AdminInterfaces } from './src/components/tools/AdminInterfaces.vue';
export { default as AdminOptions } from './src/components/tools/AdminOptions.vue';

// Export Pages
export { default as ActionPage } from './src/pages/ActionPage.vue';
export { default as DashboardPage } from './src/pages/DashboardPage.vue';
export { default as DataTablePage } from './src/pages/DataTablePage.vue';

// Export Router and Utilities
export { default as routes } from './src/router';
export {  defineAbilitiesFor } from './src/utils/ability';
export { default as FirebaseManager } from './src/FirebaseManager';
